pkg/frontend/querymiddleware/results_cache.go

```diff
@@ -639,3 +660,64 @@ func cacheHashKey(key string) string {
 	// Hex because memcache keys must be non-whitespace non-control ASCII
 	return hex.EncodeToString(hasher.Sum(nil))
 }
+
+// approximateSamplesProcessedPerStep approximates the samples processed per step from the total samples processed.
+// We had to use appoximation, becase MQE doesn't support per step stats yet.
+func approximateSamplesProcessedPerStep(start int64, end int64, step int64, samplesProcessed uint64) []StepStat {
+	numSteps := (end-start)/step + 1
+	// approxPerStep is not float becase:
+	// 1. Real per step stats always have integer values. Having int in approximation will simplify migration to per step stats from engine
+	// 2. approxPerStep < 0 doesn't make sense
+	// 3. approxPerStep is rounded with math.Ceil, because in most of the real-life scenarios, samplesProcessed / numSteps is > 1.
+	approxPerStep := int64(math.Ceil(float64(samplesProcessed) / float64(numSteps)))
+	perStepStats := make([]StepStat, 0, numSteps)
+	for i := int64(0); i < numSteps; i++ {
+		perStepStats = append(perStepStats, StepStat{
+			Timestamp: start + i*step,
+			Value:     approxPerStep,
+		})
+	}
+	return perStepStats
+}
+
+// extractSamplesProcessedPerStep extracts the per step samples count for the subrange within the extent.
+func extractSamplesProcessedPerStep(extent Extent, start int64, end int64) []StepStat {
+	// Validate the subrange is valid and within the extent.
+	if end < start || start < extent.Start || end > extent.End {
+		return nil
+	}
+
+	var result []StepStat
+	for _, step := range extent.SamplesProcessedPerStep {
+		if start <= step.Timestamp && step.Timestamp <= end {
+			result = append(result, step)
+		}
+	}
+
+	return result
+}
+
+func mergeSamplesProcessedPerStep(steps ...[]StepStat) []StepStat {
```

Comment:
If we simply define that `StepStat` must be ordered by timestamp, then you don't need to use a map here. You can just go building the merged slice in a single pass. Keep it simple, just take 2 inputs. Don't need to be variadic.